### PR TITLE
Flavioarfaria/checkout confirmation

### PR DESCRIPTION
**Pull request by flavioarfaria**
Requests full wallet, extracts token, updates email address and completes checkout.

### CODE REVIEW

**Comment by flavioarfaria** [comment]
Since `ApplePayPayment` has been refactored into `NativePayment` that *might* have some impact on iOS. Since the only changes made were basically renaming the class and adding a new member, the impact, if exists, is minimal.

File: android/shopify_buy_plugin/src/main/java/com/shopify/unity/buy/models/NativePayment.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    private static final String IDENTIFIER = "Identifier";
+    private static final String TOKEN_DATA = "TokenData";
+
+    @NonNull public final MailingAddressInput billingContact;
+    @NonNull public final ShippingContact shippingContact;
+    @Nullable public final String identifier;
+    @NonNull public final TokenData tokenData;
+
+    private NativePayment(Builder builder) {
+        billingContact = builder.billingContact;
```
Comment (sleroux) [inline android/shopify_buy_plugin/src/main/java/com/shopify/unity/buy/models/NativePayment.java]: Probably minor but what happens if we never set `billingContact` in the builder and we assign null to this non-null property? Does it make sense to add a layer of validation in the `build` method to make sure we don't try to do this?

File: scripts/generator/graphql_generator/csharp/SDK/Android/AndroidNativeCheckout.AndroidPayEventReceiver.cs.erb [hunk 2]
```diff
@@ -32,6 +35,49 @@ namespace <%= namespace %>.SDK.Android {
...
+                NativeMessage message) {
+            CartState.CheckoutWithTokenizedPayment(tokenizedPaymentInput, (ShopifyError error) => {
+                AndroidPayCheckoutResponse.Status status = error == null
+                        ? AndroidPayCheckoutResponse.Status.Success
+                        : AndroidPayCheckoutResponse.Status.Failure;
+                message.Respond(new AndroidPayCheckoutResponse(status).ToJsonString());
+            });
+        }
+
+        private void SetFinalCheckoutFieldsForPayment(NativePayment payment, Checkout checkout,
```
Comment (sleroux) [inline scripts/generator/graphql_generator/csharp/SDK/Android/AndroidNativeCheckout.AndroidPayEventReceiver.cs.erb]: I don't see this being called anywhere. Shouldn't this be called just before performing the checkout?

**Comment by flavioarfaria** [comment]
@sleroux regarding `SetFinalCheckoutFieldsForPayment()`, here's what it does on iOS:
1. Delegates to `CartState.SetFinalCheckoutFields()` (more on that later);
1. Tries to recover from user errors in its callback, such as invalid address information.

    There's not much to try to recover on Android Pay, since Android Pay itself validates address information in its own address form.

Regarding `CartState.SetFinalCheckoutFields()` (mentioned above in 1), it performs three mutations:

1. Email update;

    We're doing this in `OnConfirmCheckout()`.

1. Shipping address update;

    We're doing this whenever the user explicitly changes the address in the confirmation screen. That's why I needed to be able to send messages to Unity from an activity other than Unity's and that's why that single-activity approach came into play.

1. Shipping rate update.

    We're also doing this whenever the user explicitly does so.

To wrap up, it turns out I found `SetFinalCheckoutFieldsForPayment()` does a bunch of stuff we've done previously and we don't need to do it over (as far as I understand). The only thing it does we actually need to do is to update the email address, and we're calling it directly instead.

For 2. and 3., we're also setting defaults right after we get the masked wallet. The default shipping rate being set is the first one in the array. You can see this being done in `OnUpdateShippingAddress()` and `RespondToUpdateAddressSuccessForMessage()` respectively.

Let me know if I'm missing (or misunderstanding) something.

File: android/shopify_buy_plugin/src/main/java/com/shopify/unity/buy/models/NativePayment.java (hunk 1 above)
Comment (flavioarfaria) [reply android/shopify_buy_plugin/src/main/java/com/shopify/unity/buy/models/NativePayment.java]: Of course it does. Will do it. :)

File: scripts/generator/graphql_generator/csharp/SDK/Android/AndroidNativeCheckout.AndroidPayEventReceiver.cs.erb (hunk 2 above)
Comment (flavioarfaria) [reply scripts/generator/graphql_generator/csharp/SDK/Android/AndroidNativeCheckout.AndroidPayEventReceiver.cs.erb]: This is garbage and must be removed.

Comment (sleroux) [reply scripts/generator/graphql_generator/csharp/SDK/Android/AndroidNativeCheckout.AndroidPayEventReceiver.cs.erb]: Ah okay. Regarding your comment below, having everything be set at each point in the flow is the best approach. I just saw this method here and no references being made to it. All good 👍

Comment (flavioarfaria) [reply scripts/generator/graphql_generator/csharp/SDK/Android/AndroidNativeCheckout.AndroidPayEventReceiver.cs.erb]: https://github.com/Shopify/unity-buy-sdk/pull/356/commits/4cc425f48b0758cf55b7724c2e4057b3b63f7db9

File: android/shopify_buy_plugin/src/main/java/com/shopify/unity/buy/models/NativePayment.java (hunk 1 above)
Comment (flavioarfaria) [reply android/shopify_buy_plugin/src/main/java/com/shopify/unity/buy/models/NativePayment.java]: https://github.com/Shopify/unity-buy-sdk/pull/356/commits/3b721d459bbdfcf562a5a2f0900809b1796330fd

**Pull request merged**